<a href="https://colab.research.google.com/github/amitgit2510/Info_miners_2/blob/Nisarg/DM_CP02.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import pandas as pd

In [ ]:
!unzip "/content/drive/MyDrive/Data Science/Sem-3/Data Mining/cp02_data.zip" -d "/content/"

Archive:  /content/drive/MyDrive/Data Science/Sem-3/Data Mining/cp02_data.zip
  inflating: /content/hospital-inpatient-discharges-sparcs-de-identified-2010-1.csv  


In [ ]:
df = pd.read_csv("/content/hospital-inpatient-discharges-sparcs-de-identified-2010-1.csv")
df.head()

<ipython-input-4-7065ed1eb75f>:1: DtypeWarning: Columns (1,2,7,11,29) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv("/content/hospital-inpatient-discharges-sparcs-de-identified-2010-1.csv")


index Health Service Area Hospital County  Operating Certificate Number  \
0      0      Capital/Adiron          Albany                      101000.0   
1      1      Capital/Adiron          Albany                      101000.0   
2      2      Capital/Adiron          Albany                      101000.0   
3      3      Capital/Adiron          Albany                      101000.0   
4      4      Capital/Adiron          Albany                      101000.0   

   Facility ID                   Facility Name    Age Group  \
0          1.0  Albany Medical Center Hospital     18 to 29   
1          1.0  Albany Medical Center Hospital     50 to 69   
2          1.0  Albany Medical Center Hospital  70 or Older   
3          1.0  Albany Medical Center Hospital  70 or Older   
4          1.0  Albany Medical Center Hospital  70 or Older   

  Zip Code - 3 digits Gender   Race  ... Source of Payment 2  \
0                 NaN      F  White  ...            Self-Pay   
1                 NaN      M  White  ...            Medicare   
2                 NaN      M  White  ...          Blue Cross   
3                 NaN      F  White  ...   Insurance Company   
4                 NaN      F  White  ...   Insurance Company   

  Source of Payment 3 Attending Provider License Number  \
0                 NaN                            3623.0   
1          Blue Cross                          216951.0   
2            Self-Pay                            3076.0   
3            Self-Pay                          140796.0   
4            Self-Pay                          170254.0   

  Operating Provider License Number  Other Provider License Number  \
0                               NaN                            NaN   
1                          216951.0                            NaN   
2                               NaN                            NaN   
3                          140796.0                       170254.0   
4                          170254.0                       170254.0   

   Birth Weight Abortion Edit Indicator  Emergency Department Indicator  \
0             0                       N                               Y   
1             0                       N                               Y   
2             0                       N                               N   
3             0                       N                               N   
4             0                       N                               N   

  Total Charges  Total Costs  
0       4476.23      1672.65  
1     148612.34     51414.70  
2      16561.99      4032.49  
3      13593.51      4947.81  
4      31962.58     16988.33  

[5 rows x 38 columns]

In [ ]:
df.columns

Index(['index', 'Health Service Area', 'Hospital County',
       'Operating Certificate Number', 'Facility ID', 'Facility Name',
       'Age Group', 'Zip Code - 3 digits', 'Gender', 'Race', 'Ethnicity',
       'Length of Stay', 'Type of Admission', 'Patient Disposition',
       'Discharge Year', 'CCS Diagnosis Code', 'CCS Diagnosis Description',
       'CCS Procedure Code', 'CCS Procedure Description', 'APR DRG Code',
       'APR DRG Description', 'APR MDC Code', 'APR MDC Description',
       'APR Severity of Illness Code', 'APR Severity of Illness Description',
       'APR Risk of Mortality', 'APR Medical Surgical Description',
       'Source of Payment 1', 'Source of Payment 2', 'Source of Payment 3',
       'Attending Provider License Number',
       'Operating Provider License Number', 'Other Provider License Number',
       'Birth Weight', 'Abortion Edit Indicator',
       'Emergency Department Indicator', 'Total Charges', 'Total Costs'],
      dtype='object')

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2622133 entries, 0 to 2622132
Data columns (total 38 columns):
 #   Column                               Dtype  
---  ------                               -----  
 0   index                                int64  
 1   Health Service Area                  object 
 2   Hospital County                      object 
 3   Operating Certificate Number         float64
 4   Facility ID                          float64
 5   Facility Name                        object 
 6   Age Group                            object 
 7   Zip Code - 3 digits                  object 
 8   Gender                               object 
 9   Race                                 object 
 10  Ethnicity                            object 
 11  Length of Stay                       object 
 12  Type of Admission                    object 
 13  Patient Disposition                  object 
 14  Discharge Year                       int64  
 15  CCS Diagnosis Code              

In [ ]:
tmp = 100*df.isnull().sum() / len(df)
pd.DataFrame(tmp.sort_values(ascending=False), columns = ['Null values %'])

Null values %
Other Provider License Number            81.383782
Source of Payment 3                      77.831330
Source of Payment 2                      30.905412
Operating Provider License Number        25.582913
Attending Provider License Number         0.186375
Hospital County                           0.186375
Operating Certificate Number              0.186375
Facility ID                               0.186375
Health Service Area                       0.186375
Zip Code - 3 digits                       0.122076
CCS Procedure Description                 0.086685
CCS Procedure Code                        0.086685
CCS Diagnosis Description                 0.086685
CCS Diagnosis Code                        0.086685
APR Risk of Mortality                     0.009191
APR Severity of Illness Description       0.009191
Patient Disposition                       0.003928
Birth Weight                              0.000000
Abortion Edit Indicator                   0.000000
Emergency Department Indicator            0.000000
Total Charges                             0.000000
Source of Payment 1                       0.000000
APR Medical Surgical Description          0.000000
index                                     0.000000
APR DRG Code                              0.000000
APR Severity of Illness Code              0.000000
APR MDC Description                       0.000000
APR MDC Code                              0.000000
APR DRG Description                       0.000000
Discharge Year                            0.000000
Type of Admission                         0.000000
Length of Stay                            0.000000
Ethnicity                                 0.000000
Race                                      0.000000
Gender                                    0.000000
Age Group                                 0.000000
Facility Name                             0.000000
Total Costs                               0.000000

Classification : Severity of illness (on basis of health service area, gender, and diagnosis).

In [ ]:
#'APR Severity of Illness Code', 'APR Severity of Illness Description'
df['APR Severity of Illness Description'].unique()

array(['Minor', 'Major', 'Moderate', 'Extreme', nan], dtype=object)

The null values in severity of illness and the selected independent variables is < 1%. Hence, we can drop the null values.

In [ ]:
df1 = df[['Health Service Area', 'Gender', 'CCS Diagnosis Code', 'APR Severity of Illness Description']].dropna()
df1.columns = ['area', 'gender', 'diagnosis', 'severity_of_illness']
df1.head()

area gender  diagnosis severity_of_illness
0  Capital/Adiron      F      135.0               Minor
1  Capital/Adiron      M      231.0               Major
2  Capital/Adiron      M       42.0               Major
3  Capital/Adiron      F      254.0            Moderate
4  Capital/Adiron      F      203.0               Minor

In [ ]:
df1['gender'].replace(to_replace='U', value='F', inplace=True)

In [ ]:
df1.describe(include='object')

area   gender severity_of_illness
count         2614735  2614735             2614735
unique              8        2                   4
top     New York City        F               Minor
freq          1232772  1471514              970896

In [ ]:
#Print percentage of value counts
for col in df1.columns:
  print('Column:', col)
  print(100 * df1[col].value_counts() / len(df1))
  print('-'*40)

Column: area
New York City     47.147111
Long Island       14.729791
Hudson Valley     10.440446
Western NY         7.059147
Capital/Adiron     6.831820
Central NY         6.533014
Finger Lakes       6.061723
Southern Tier      1.196947
Name: area, dtype: float64
----------------------------------------
Column: gender
F    56.277749
M    43.722251
Name: gender, dtype: float64
----------------------------------------
Column: diagnosis
218.0    9.217033
657.0    2.411372
108.0    2.380318
2.0      2.332015
102.0    2.296638
           ...   
53.0     0.001606
86.0     0.001606
223.0    0.001224
206.0    0.000803
174.0    0.000497
Name: diagnosis, Length: 260, dtype: float64
----------------------------------------
Column: severity_of_illness
Minor       37.131717
Moderate    36.717947
Major       20.241210
Extreme      5.909127
Name: severity_of_illness, dtype: float64
----------------------------------------


In [ ]:
#applying one-hot encoding to area, gender, race and ethnicity columns
df_encoded = pd.get_dummies(df1, columns=['area', 'gender'])

In [ ]:
df_encoded.head()

diagnosis severity_of_illness  area_Capital/Adiron  area_Central NY  \
0      135.0               Minor                    1                0   
1      231.0               Major                    1                0   
2       42.0               Major                    1                0   
3      254.0            Moderate                    1                0   
4      203.0               Minor                    1                0   

   area_Finger Lakes  area_Hudson Valley  area_Long Island  \
0                  0                   0                 0   
1                  0                   0                 0   
2                  0                   0                 0   
3                  0                   0                 0   
4                  0                   0                 0   

   area_New York City  area_Southern Tier  area_Western NY  gender_F  gender_M  
0                   0                   0                0         1         0  
1                   0                   0                0         0         1  
2                   0                   0                0         0         1  
3                   0                   0                0         1         0  
4                   0                   0                0         1         0

In [ ]:
#ordinal encoding severity of illness
custom_mapping = {
    'Minor': 1,
    'Moderate': 2,
    'Major': 3,
    'Extreme':4
}

# Use the map() function to apply the custom mapping
df_encoded['severity'] = df_encoded['severity_of_illness'].map(custom_mapping)

# Display the DataFrame with the custom ordinal encoding
df_encoded.head()

diagnosis severity_of_illness  area_Capital/Adiron  area_Central NY  \
0      135.0               Minor                    1                0   
1      231.0               Major                    1                0   
2       42.0               Major                    1                0   
3      254.0            Moderate                    1                0   
4      203.0               Minor                    1                0   

   area_Finger Lakes  area_Hudson Valley  area_Long Island  \
0                  0                   0                 0   
1                  0                   0                 0   
2                  0                   0                 0   
3                  0                   0                 0   
4                  0                   0                 0   

   area_New York City  area_Southern Tier  area_Western NY  gender_F  \
0                   0                   0                0         1   
1                   0                   0                0         0   
2                   0                   0                0         0   
3                   0                   0                0         1   
4                   0                   0                0         1   

   gender_M  severity  
0         0         1  
1         1         3  
2         1         3  
3         0         2  
4         0         1

In [ ]:
from sklearn.preprocessing import LabelEncoder

# Initialize the LabelEncoder
label_encoder = LabelEncoder()

# Fit and transform the categorical column to encode the labels
df_encoded['diagnosis_encoded'] = label_encoder.fit_transform(df_encoded['diagnosis'])

In [ ]:
df_final = df_encoded.drop(['severity_of_illness'], axis=1)
df_final.head()

diagnosis  area_Capital/Adiron  area_Central NY  area_Finger Lakes  \
0      135.0                    1                0                  0   
1      231.0                    1                0                  0   
2       42.0                    1                0                  0   
3      254.0                    1                0                  0   
4      203.0                    1                0                  0   

   area_Hudson Valley  area_Long Island  area_New York City  \
0                   0                 0                   0   
1                   0                 0                   0   
2                   0                 0                   0   
3                   0                 0                   0   
4                   0                 0                   0   

   area_Southern Tier  area_Western NY  gender_F  gender_M  severity  \
0                   0                0         1         0         1   
1                   0                0         0         1         3   
2                   0                0         0         1         3   
3                   0                0         1         0         2   
4                   0                0         1         0         1   

   diagnosis_encoded  
0                123  
1                216  
2                 41  
3                239  
4                188

In [ ]:
from sklearn.model_selection import train_test_split

In [ ]:
X = df_final.drop(['severity', 'diagnosis'], axis=1)
y = df_final['severity']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, shuffle=True)

We shall use ordinal logistic regression, logistic regression, and random forest classifier

In [ ]:
#!pip install mord

  Preparing metadata (setup.py) ... done
  Created wheel for mord: filename=mord-0.7-py3-none-any.whl size=9885 sha256=8218c8ba2b6e71b73e2992bcb161c7dcafe9ed5401132d3f42a0f2c68a2f57f5
  Stored in directory: /root/.cache/pip/wheels/77/00/19/3cea86fbfc737ec4acb515cd94497dcc33f943fa157548b96c
Successfully built mord


Ordinal Logistic Regression

In [ ]:
from mord import LogisticAT

model = LogisticAT(alpha=0.01)
model.fit(X_train[:10000], y_train[:10000])

LogisticAT(alpha=0.01)

In [ ]:
y_pred = model.predict(X_test)

In [ ]:
y_pred

array([1, 2, 2, ..., 2, 2, 2])

In [ ]:
y_test

2227269    2
1509246    3
861574     2
1480445    1
1200956    2
          ..
199578     2
1364513    1
1456694    1
2386047    3
2468676    4
Name: severity, Length: 653684, dtype: int64

In [ ]:
from sklearn.metrics import classification_report
print(classification_report(y_test, y_pred, zero_division=0))

              precision    recall  f1-score   support

           1       0.31      0.09      0.14    242289
           2       0.35      0.84      0.49    240258
           3       0.38      0.01      0.03    132416
           4       0.00      0.00      0.00     38721

    accuracy                           0.35    653684
   macro avg       0.26      0.24      0.16    653684
weighted avg       0.32      0.35      0.24    653684



Let us perform search for hyperparameter alpha:

In [ ]:
accs = []
for alpha in [0, 0.001, 0.01, 1]:
  model = LogisticAT(alpha=alpha)
  model.fit(X_train[:10000], y_train[:10000])

  y_pred = model.predict(X_train[10000:10500])
  y_true = y_train[10000:10500]

  acc = (y_true==y_pred).astype(int).sum() / 500
  accs.append(acc)

accs

[0.338, 0.338, 0.338, 0.338]

As we can see, none of the combinations produces good accuracy


Let us try logistic regression

In [ ]:
from sklearn.linear_model import LogisticRegression

accs = []
for alpha in [0.001, 0.01, 1, 10]:
  model = LogisticRegression(penalty='l2', C=alpha)
  model.fit(X_train[:10000], y_train[:10000])

  y_pred = model.predict(X_train[10000:10500])
  y_true = y_train[10000:10500]

  acc = (y_true==y_pred).astype(int).sum() / 500
  accs.append(acc)

accs

/usr/local/lib/python3.10/dist-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(
/usr/local/lib/python3.10/dist-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _c

[0.426, 0.398, 0.4, 0.398]

Logistic regression with C=0.001 produces the best accuracy. Let us train the model with this parameter and more data.

In [ ]:
#Taking 1 Lakh values
model = LogisticRegression(penalty='l2', C=0.001)
model.fit(X_train[:100000], y_train[:100000])

y_pred = model.predict(X_test)

# Print a classification report with precision, recall, and F1-score
report = classification_report(y_test, y_pred)
print(report)

/usr/local/lib/python3.10/dist-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels wit

              precision    recall  f1-score   support

           1       0.44      0.60      0.51    242289
           2       0.37      0.46      0.41    240258
           3       0.34      0.07      0.12    132416
           4       0.00      0.00      0.00     38721

    accuracy                           0.40    653684
   macro avg       0.29      0.28      0.26    653684
weighted avg       0.37      0.40      0.36    653684



/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


We get an accuracy of 40%

Finally, let us try Random Forest Classifier:

In [ ]:
from sklearn.ensemble import RandomForestClassifier
rf_classifier = RandomForestClassifier(n_estimators=100, random_state=42)

# Fit the model on the training data
rf_classifier.fit(X_train[:10000], y_train[:10000])

# Make predictions on the test data
y_pred = rf_classifier.predict(X_test)

# Print a classification report with precision, recall, and F1-score
report = classification_report(y_test, y_pred)
print(report)

              precision    recall  f1-score   support

           1       0.56      0.59      0.57    242289
           2       0.45      0.47      0.46    240258
           3       0.36      0.34      0.35    132416
           4       0.35      0.18      0.24     38721

    accuracy                           0.47    653684
   macro avg       0.43      0.40      0.41    653684
weighted avg       0.47      0.47      0.47    653684



Let us perform search over hyperparameter n_estimators:

In [ ]:
from sklearn.linear_model import LogisticRegression

accs = []
for n in [25, 50, 75, 100, 200]:
  rf_classifier = RandomForestClassifier(n_estimators=n, random_state=42)

  # Fit the model on the training data
  rf_classifier.fit(X_train[:10000], y_train[:10000])

  # Make predictions on the test data
  y_pred = rf_classifier.predict(X_train[10000:10500])
  y_true = y_train[10000:10500]

  acc = (y_true==y_pred).astype(int).sum() / 500
  accs.append(acc)

accs

[0.446, 0.456, 0.452, 0.448, 0.452]

Best accuracy is produced by n_estimators = 75. Let us train the model with more data.

In [ ]:
rf_classifier = RandomForestClassifier(n_estimators=75, random_state=42)

rf_classifier.fit(X_train[:100000], y_train[:100000])

RandomForestClassifier(n_estimators=75, random_state=42)

In [ ]:
y_pred = rf_classifier.predict(X_test)
print(classification_report(y_test, y_pred, zero_division=0))

              precision    recall  f1-score   support

           1       0.60      0.64      0.62    242289
           2       0.46      0.55      0.50    240258
           3       0.42      0.27      0.33    132416
           4       0.49      0.24      0.32     38721

    accuracy                           0.51    653684
   macro avg       0.49      0.42      0.44    653684
weighted avg       0.50      0.51      0.50    653684



We get an accuracy of 51% on the test set.